[DataType] Add bfloat16

Menooker opened the pull request:
We add bfloat16 as a new type named "bf16" in the frontend. Completed LLVM backend for generating bf16.

 * Use int16 as the storage type in LLVM
 * Add legalization to enable computations on bf16
 * Add runtime frontend support (e.g. allow converting numpy's uint16 array to bf16 NDArray)

# Details on legalization

Since most of the HW has no native support for computation on bf16, we added a pass `BF16Legalization` to use fp32 computing bf16 data. It adds `cast_to_fp32()` before each Op involing bf16 operands, and use Ops of fp32 to compute. Finally, it adds a 'cast_to_bf16()' after each Op that is altered. e.g.

`add(a,b)` => `cast16(add(cast32(a), cast32(b)))`

We call this phase as "BF16Promotion". It is a sub-pass of `BF16Legalization` pass.

We note that this will add redundant casting. e.g.

`add(a, neg(b))` => `cast16(add(cast32(a), cast32(cast16(neg(cast32(b))))) `

The pattern `cast32(cast16(some_fp32_value))` can be simplified to `some_fp32_value`.

Thus, we add an optimization pass after "BF16Promotion" in `BF16Legalization` pass, which eliminates redundant casts.

After `BF16Legalization` pass, there will be no bf16 related computation in the AST, except casting between fp32 and bf16, bf16 value comparasion and assignment.

# Casting between fp32 and bf16

We follow PyTorch's bf16 [casting](https://github.com/pytorch/pytorch/blob/master/c10/util/BFloat16.h) implementation.

Menooker commented:
@zhiics @liangfu Please help review this PR. Thanks!

tqchen commented:
cc @gussmith23 might be related to BYOD

liangfu commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -309,6 +309,9 @@ llvm::Type* CodeGenLLVM::DTypeToLLVMType(const DataType& dtype) const {
         LOG(FATAL) << "do not support " << dtype;
     }
+  } else if (dtype.is_bfloat()) {
+    CHECK_EQ(dtype.bits(), 16);
```
Since bfloat is assumed to be 16bit, can we keep the terminology more consistent? Since the data type is termed as `bf`, `bf16`, `bfloat16`, `bfloat` in the proposed change. Or are we going to support more data types like bfloat18 and bfloat20 in the future?

Menooker replied:
Sorry for the inclarity. I think in bfloat[X], only X=16 makes sense. But TVM's type system allows specifying the `bits` of a type. So here is the checking to make sure it is bf16.

ZhennanQin replied:
A good question. Will we treat TensorFloat-32 as bfloat20? If so, then `bits` is useful to distinguish those.

liangfu commented on tests/python/unittest/test_tir_transform_bf16_legalize.py:
```diff
@@ -0,0 +1,139 @@
+if __name__ == "__main__":
+    test_promote()
+    test_eliminate()
+    test_legalize()
```
Please leave a new line at EOF, even this is test script :)

liangfu commented on tests/python/unittest/test_target_codegen_llvm.py:
```diff
@@ -710,6 +710,52 @@ def _transform(f, *_):
+def np_bf162np_float(arr):
+    ''' Convert a numpy array of bf16 (uint16) to a numpy array 
+    of float'''
+    u32 = np.left_shift(arr.astype('uint32'), 16)
```
Are we going to produce a potential endianness problem here?

Menooker replied:
In my understanding, fp32=>bf16 casting preserves the higher-ordered bits (bits 31-16). We don't need to know whether the higher-ordered bits are stored in a larger address or a smaller address (which is the endianness), we just need to get the bits by shifting, which is well-defined - just using shifting is enough.  

Reference: [wiki for fp32 bit order](https://en.wikipedia.org/wiki/Single-precision_floating-point_format)

[PyTorch's bf16 casting] (https://github.com/pytorch/pytorch/blob/master/c10/util/BFloat16.h)

Menooker replied:
I am not 100% sure about this. I have tested the code on x86, not (yet) on other arch.

liangfu replied:
Can we reused the following code snippet, which preserves endianness checks?

https://github.com/apache/incubator-tvm/blob/6cbda80227fc18a859c4b01f57f75abbd7a16181/3rdparty/bfloat16/bfloat16.cc#L27-L35

Menooker replied:
If my understanding is correct, we don't need to care about endianness. BF16 conversions only involves getting higher-ordered bits. And the operation to get higher-ordered bits in C++/Numpy is well-defined.

liangfu commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -906,7 +954,7 @@ DEFINE_CODEGEN_BINARY_OP(Mul);
     if (t.is_int()) {                                                                \
       return builder_->CreateICmpS##Op(a, b);                                        \
-    } else if (t.is_uint()) {                                                        \
+    } else if (t.is_uint() || t.is_bfloat()) {                                       \
```
Isn't comparing bfloat16 this way risky?

Menooker replied:
FP32/FP64 comparasion are also bit-wise in my understanding.

liangfu commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -555,12 +558,48 @@ void CodeGenLLVM::CreateSerialFor(llvm::Value* begin, llvm::Value* end, llvm::Va
+                                    ? static_cast<llvm::Type*>(builder_->getInt32Ty())
+                                    : llvm::VectorType::get(builder_->getInt32Ty(), from.lanes());
+    auto v = builder_->CreateZExt(value, extended_type);
+    v = builder_->CreateShl(v, 16);
```
Potential endianness problem here?

tqchen commented on include/tvm/runtime/c_runtime_api.h:
```diff
@@ -114,6 +114,7 @@ typedef enum {
   kTVMNNVMLast = 20U,
   // The following section of code is used for non-reserved types.
   kTVMExtReserveEnd = 64U,
+  kTVMBFloat = 65U,
```
We do not want BFloat to be passed as PackedFunc argument, most packedfunc argument should always be passed as double

Menooker replied:
I suppose TVM should support kernel generation, e.g. generating a fused "conv+bn+relu", rather than generating end-to-end model, which is the usual case. In this case, we might select some intermediate layers of the model and let TVM generate the selected layers. The layers may require bf16 as the dtype, as they are in the middle of the model.

What I want to say is that we sometimes need bf16 as the input dtype.  In our usecase in Intel, we need to generate a bf16 kernel (e.g. conv+bn+relu).

tqchen replied:
Such dtype is covered by allocating a DLTensor with type_code equals kBFloat, and does not need patch to the code here(needed for parameter argument passing PackedFunc).

tqchen replied:
The particular code is used when we directly pass a constant into PackedFunc, e.g. `f(1.0, some_float_value)`. in these cases double can be used.

Menooker replied:
If we remove this type from TVM runtime, we cannot pass a bf16 array to TVM via Python and users can only pass bf16 buffers via C runtime (or in some awkward way to construct a bf16 DLTensor via Python). Currently, with  kTVMBFloat defined, we can:

```python
        A = te.placeholder((32, ), dtype='bfloat16')
        B = te.placeholder((32, ), dtype='bfloat16')
        d = te.compute((32, ), lambda x: A[x] + B[x])
        sch = te.create_schedule(d.op)
        module = tvm.build(sch, [A, B, d])
        npa = np.random.rand(32).astype('float32')
        npb = np.random.rand(32).astype('float32')
        a_ = np_float2tvm_bf16(npa)
        b_ = np_float2tvm_bf16(npb)
        c_ = tvm.nd.empty((32,), 'bfloat16')
        module(a_, b_, c_)
```

Which is useful for testing and prototyping.

tqchen replied:
I don't think you will `kTVMBFloat` to support this feature. The kDLBFloat flag in the runtime::DataType should be sufficient for NDArray contents(because the type code in the NDArray contents diverges from the TVM type code above the OpaqueHandle).

Menooker replied:
ok I understand. will change that

tqchen commented on include/tvm/runtime/data_type.h:
```diff
@@ -81,6 +82,10 @@ class DataType {
   /*! \return whether type is a float16 type. */
   bool is_float16() const { return is_float() && bits() == 16; }
+  /*! \return whether type is a bfloat type. */
+  bool is_bfloat() const { return code() == DataType::kBFloat; }
```
given that only bfloat16 is defined, is_bf16 is a good enough function

Menooker replied:
ok, changed

tqchen commented on include/tvm/runtime/data_type.h:
```diff
@@ -297,6 +302,8 @@ inline const char* TypeCode2Str(int type_code) {
     case kTVMObjectRValueRefArg:
       return "ObjectRValueRefArg";
+    case kTVMBFloat:
+      return "bf";
```
bfloat

Menooker replied:
ok, changed

Menooker replied:
ok, changed

tqchen commented on src/target/llvm/codegen_llvm.cc:
```diff
@@ -555,12 +558,48 @@ void CodeGenLLVM::CreateSerialFor(llvm::Value* begin, llvm::Value* end, llvm::Va
   if (to.is_handle()) {
     return builder_->CreateBitCast(value, target);
+  } else if (to.is_float() && from.is_bfloat()) {
+    CHECK_EQ(from.bits(), 16);
```
If LLVM does not support bfloat, then perhaps we should do the legalization as a TIR=>TIR pass as opposed to do it in LLVM

Menooker replied:
We are actually doing TIR=>TIR legalization pass in TVM. See src/tir/transforms/bf16_legalize.cc

tqchen replied:
Then we should directly change the type to be i16 during legalization and remove special handling code for bfloat16

Menooker replied:
Then we cannot tell whether it is a float32 => i16 or float32 => bfloat16 casting

ZhennanQin replied:
There're 2 kinds of legalization:
1. TIR->TIR. TIR has full ability to describe any bfloat16 operation after this PR. This legalization is introduced just because of hardware limitation that current hardware only provide few bfloat16 operations. One day when hardware has full instructions support with bfloat16, ideally this legalization can be skipped. So this legalization is a target dependent pass. 
2. TIR->LLVM IR. I guess this is the legalization that @tqchen mentions. Because LLVM IR doesn't natively support bfloat16 , i16 will be used to replace bfloat16. In this PR, I guess this is done within codegen_llvm, not by a particular pass.

tqchen replied:
Then we should legalize the cast as well in the TIR to introduce additional cast funtions, please also refer to https://tvm.apache.org/2020/05/20/bring-your-own-datatypes for releated implemenetation

Menooker replied:
Just 2 small questions.

Did you mean totally eliminating bf16 dtype in legalization pass? This will bring much more complexity in the BF16Legalize pass, because we need to check every TIR node to replace bf16 with int16. In contrast, current impl only changes computation TIR nodes. And in the codegen, the bf16 generation is quite simple, just adding another ‘else if’ in casting node and tvm dtype to llvm type converter

Menooker replied:
and I think the way processing “custom data type” that you mentioned does not fit this pr well. Actually I have already notice this feature before I wrote this bf16 feature. But it needs function calls to do lowering, which is not friendly to the codegen backend to do auto vectorization and so on. Of course you can say we can implement this cast function as an intrinsic. Yes, but more complexity is brought. 

I think letting bf16 dtype live until codegen is a good idea, it makes legalization, impl of casting easier

tqchen commented:
Given that this is a new feature that will affect quite some people, please open a new RFC thread in the discuss forum to describe the motivation and the high level design. Thank you!

Menooker commented:
@tqchen I have changed the legalization pass to pure TIR->TIR. Added a pass called `BF16Lowering` to replace bf16 type with uint16 and lower the casting by shifting.

I have met a problem that I need to replace `BufferNode` of bf16 with uint16. It works well and can produce same result as numpy. However, a lowered bf16 cast cannot pass `assert_structural_equal` test, like [here](https://github.com/Menooker/incubator-tvm/blob/177f99a8ce7f95e927cabeaf18684ade0c8ebc73/tests/python/unittest/test_tir_transform_bf16_legalize.py#L143). Is this a problem of this PR?

tqchen commented:
@liangfu please followup with the reviews. @gussmith23 it would be great if you can also help to take a look

liangfu commented on src/tir/transforms/bf16_legalize.cc:
```diff
@@ -0,0 +1,384 @@
+      auto uint32_v =
+          CallNode::make(uint32_dtype, CallNode::reinterpret, {op_val}, CallNode::PureIntrinsic);
+      auto uint16_dtype = DataType(kDLUInt, 16, op_val->dtype.lanes());
+      // uint32_t rounding_bias = ((U32 >> 16) & 1) + UINT32_C(0x7FFF);
```
clean up

Menooker replied:
can we keep this line? I left it here to illustrate what the following lines do.

ZhennanQin replied:
@Menooker Then you'd better re-format this line as a comment. Like line 120-128

liangfu commented on src/tir/transforms/bf16_legalize.cc:
```diff
@@ -0,0 +1,384 @@
+      CHECK(op->dtype.is_float() && op->dtype.bits() == 32);
+      auto uint32_dtype = DataType(kDLUInt, 32, op_val->dtype.lanes());
+      auto uint32_v = CastNode::make(uint32_dtype, op_val);
+      return CallNode::make(op->dtype, CallNode::reinterpret, {uint32_v << 16},
```
add a note for future update in endian invariant like 

https://github.com/apache/incubator-tvm/blob/95b3ad971dfed6aa4e9f18d4f43c846806fd5292/src/runtime/stackvm/stackvm.cc#L211

liangfu commented on src/tir/transforms/bf16_legalize.cc:
```diff
@@ -0,0 +1,384 @@
+      // uint32_t rounding_bias = ((U32 >> 16) & 1) + UINT32_C(0x7FFF);
+      // return static_cast<uint16_t>((U32 + rounding_bias) >> 16);
+      auto rounding_bias = ((uint32_v >> 16) & 1) + make_const(uint16_dtype, 0x7FFF);
+      return CastNode::make(uint16_dtype, {(uint32_v + rounding_bias) >> 16});
```
add a note for future update in endian invariant like 

https://github.com/apache/incubator-tvm/blob/95b3ad971dfed6aa4e9f18d4f43c846806fd5292/src/runtime/stackvm/stackvm.cc#L211

Menooker replied:
ok, done

Menooker commented:
RFC Discussion link
https://discuss.tvm.ai/t/rfc-add-bfloat16-data-type/6778

tqchen commented on include/tvm/runtime/c_runtime_api.h:
```diff
@@ -114,6 +114,7 @@ typedef enum {
   kTVMNNVMLast = 20U,
   // The following section of code is used for non-reserved types.
   kTVMExtReserveEnd = 64U,
+  kTVMBFloat = 65U,
```
plase remove the flag

tqchen commented on include/tvm/runtime/data_type.h:
```diff
@@ -46,6 +46,7 @@ class DataType {
     kUInt = kDLUInt,
     kFloat = kDLFloat,
     kHandle = TVMTypeCode::kTVMOpaqueHandle,
+    kBFloat = kTVMBFloat,
```
Use kDLBfloat as it has already been accepted to the latest dlpack

tqchen commented on python/tvm/_ffi/runtime_ctypes.py:
```diff
@@ -58,7 +58,8 @@ class DataType(ctypes.Structure):
         2 : 'float',
-        4 : 'handle'
+        4 : 'handle',
+        65: 'bfloat'
```
use a code that is consistent with dlpack

Menooker replied:
@tqchen Hi, I found an inconsistency here: in Python binding, 4 is mapped to 'handle', while in C++ code, 4 is for Nullptr and 3 is for handle. Is this expected or a bug?

tqchen replied:
The definition of DataType code and PackedFunc argument type code starts to diverge after 3

Menooker replied:
May I ask what I am supposed to do here? In the newest commit, I changed `4 : 'handle'` here to 3, and let 4 mapped to bfloat. I also changed `class TypeCode(object):` and relavent C++ class to let `NULL` mapped to 21, instead of 4. I have noticed that in `class TypeCode(object):`, we have to define both `BFLOAT` and `NULL`, so I thought we needed to resolve the conflict as they were all mapped to 4.

Menooker replied:
4 is the type code of bfloat in dlpack

tqchen commented on src/tir/transforms/bf16_legalize.cc:
```diff
@@ -0,0 +1,387 @@
+    return StmtExprMutator::VisitExpr_(op);
+  }
+
+  void alter_buffers(PrimFuncNode* op) {
```
Code style, use CamelCase as per Google C Style

Menooker commented:
Some changes to the type code: We now use kDLBfloat defined in DLPack, which equals to 4. It has a conflict with old type code `kTVMNullptr = 4U`, so I move it to `kTVMNullptr = 21U'

tqchen commented on python/tvm/_ffi/_cython/base.pxi:
```diff
@@ -27,7 +27,7 @@ cdef enum TVMTypeCode:
     kFloat = 2
     kTVMOpaqueHandle = 3
-    kTVMNullptr = 4
+    kBFloat = 4
```
sorry, we want to keep the type code for the normal argument as it is, and not changin the FFI

junrushao1994 replied:
shall we remove this?

Menooker replied:
changed

tqchen commented on include/tvm/runtime/data_type.h:
```diff
@@ -372,7 +372,7 @@ inline DLDataType String2DLDataType(std::string s) {
     return t;
   } else if (s.substr(0, 6) == "bfloat") {
-    t.code = kTVMBFloat;
+    t.code = kDLBfloat;
```
the type code should directly use DataType::kBfloat

junrushao1994 replied:
I agree with tq

Menooker replied:
changed

tqchen commented:
I want to clarify the case of the TypeCode. Given that we are not trying to making a consistent type code for arguments beyond kOpaqueHandle.

So in this case, we shall simply keep the 4 for kTVMNullPtr code, not changing the FFI, and separate DataTypeCode and Argument type code.

tqchen commented:
See the data type code defined here https://github.com/apache/incubator-tvm/blob/master/include/tvm/runtime/data_type.h#L48

tqchen commented:
https://github.com/apache/incubator-tvm/pull/5730 splits the two type codes, we only need to add BFLoat16 to the DataTypeCode. cc @Menooker

Menooker commented:
@tqchen Thanks for the clarification. I have changed kTVMNullPtr back to 4.

tqchen commented:
@vinx13 @ZihengJiang @liangfu it would be great if you cam take another look. Thanks @Menooker for keep improving the PR

tqchen commented:
@junrushao1994 can you also take a quick look at this PR. thank you!

junrushao1994 commented on tests/python/unittest/test_tir_transform_bf16_legalize.py:
```diff
@@ -0,0 +1,152 @@
+        func = tvm.driver.build_module.form_irmodule(s, [a,b,c], "main", None)["main"]
+        tvm.ir.assert_structural_equal(stmt, func.body)
+
+    def orig1(a,b):
+        return lambda i: a[i]+b[i]+a[99-i]+b[99-i]
+    def after1(a,b):
+        return lambda i: to16(to32(a[i])+to32(b[i])+to32(a[99-i])+to32(b[99-i]))
+    def orig2(a,b):
+        return lambda i: a[i]*b[i]+a[99-i]*b[99-i]+a[i]
+    def after2(a,b):
+        return lambda i: to16(to32(a[i])*to32(b[i])+to32(a[99-i])*to32(b[99-i])+to32(a[i]))
```
i am not so why the coding style here can pass pylint...Mind sending a simple fix?

Menooker replied:
Sorry for that. Now I have formatted and manually run pylint on this test file. BTW, the test python files are never checked in TVM CI's pylint :)

junrushao1994 replied:
Ooops I see. That makes sense then :-)

tqchen commented:
@vinx13 @ZihengJiang @liangfu it would be great if you cam take another look and https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly

junrushao1994 commented on include/tvm/runtime/data_type.h:
```diff
@@ -72,6 +73,9 @@ class DataType {
     data_.bits = static_cast<uint8_t>(bits);
     data_.lanes = static_cast<uint16_t>(lanes);
+    if (code == kBFloat) {
+      CHECK_EQ(bits, 16);
```
It is understandable that right now we only support bf16, but my concern is that "should we put the check here"?

Menooker replied:
I understand your concern. Any suggestions for the location where we put this check? Thanks.

junrushao1994 replied:
@tqchen This is just a nitpick. What do you think?

tqchen replied:
let us leave it as it is for now, we can come back to it later

junrushao1994 commented on python/tvm/_ffi/runtime_ctypes.py:
```diff
@@ -96,6 +98,9 @@ def __init__(self, type_str):
             bits = 64
             head = ""
+        elif head.startswith("bfloat"):
+            self.type_code = 4
```
not sure if it is good to hard code here

Menooker replied:
> not sure if it is good to hard code here

Change to DataTypeCode. TVM refactors a lot (which is good). And when this PR is raised, all the type code here uses hard codes.

The other two issues you raised are also changed as required.

Menooker commented on src/tir/transforms/bf16_legalize.cc:
```diff
@@ -0,0 +1,387 @@
+
+// implementation from
+// https://github.com/pytorch/pytorch/blob/master/c10/util/BFloat16.h
+inline uint16_t round_to_nearest_even(float src) {
```
changed

tqchen commented:
@junrushao1994 @ZhennanQin please followup and https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly

tqchen commented:
Thanks @Menooker for being patient and keep improving the PR to maintain a high quality standard! Thanks @ZhennanQin @junrushao1994 @liangfu for helpful reviews!
